hydra-node/src/Hydra/Chain/Direct/Wallet.hs

```diff
@@ -187,56 +135,66 @@ watchUTxOUntil predicate TinyWallet{getUTxO} = atomically $ do
   u <- getUTxO
   u <$ check (predicate u)
 
-withTinyWallet ::
+-- | Create a new tiny wallet handle. This directly queries the wallet state
+-- from a cardano node listening at 'socketPath'.
+newTinyWallet ::
   -- | A tracer for logging
   Tracer IO TinyWalletLog ->
   -- | Network identifier to which we expect to connect.
   NetworkId ->
   -- | Credentials of the wallet.
   (VerificationKey PaymentKey, SigningKey PaymentKey) ->
-  -- | A cross-platform abstraction for managing I/O operations on local sockets
-  IOManager ->
   -- | Path to a domain socket used to connect to the server.
   FilePath ->
-  (TinyWallet IO -> IO a) ->
-  IO a
-withTinyWallet tracer networkId (vk, sk) iocp addr action = do
-  utxoVar <- newEmptyTMVarIO
-  tipVar <- newTVarIO genesisPoint
-  res <-
-    race
-      (action $ newTinyWallet utxoVar)
-      ( connectTo
-          (localSnocket iocp)
-          nullConnectTracers
-          (versions networkId $ client tracer tipVar utxoVar address)
-          addr
-      )
-  case res of
-    Left a -> pure a
-    Right () -> error "'connectTo' cannot gracefully terminate but did?"
- where
-  address =
-    toLedgerAddr $
-      mkVkAddress @Cardano.Api.Era networkId vk
-
-  newTinyWallet utxoVar =
+  IO (TinyWallet IO)
+newTinyWallet tracer networkId (vk, sk) socketPath = do
+  utxoVar <- newTVarIO =<< queryUTxOEtc
+  pure
     TinyWallet
       { getUTxO =
-          (\(u, _, _, _) -> u) <$> readTMVar utxoVar
-      , getAddress =
-          address
+          (\(u, _, _, _) -> u) <$> readTVar utxoVar
       , sign = Util.signWith (vk, sk)
       , coverFee = \lookupUTxO partialTx -> do
-          (walletUTxO, pparams, systemStart, epochInfo) <- readTMVar utxoVar
+          (walletUTxO, pparams, systemStart, epochInfo) <- readTVar utxoVar
           case coverFee_ pparams systemStart epochInfo lookupUTxO walletUTxO partialTx of
             Left e ->
               pure (Left e)
             Right (walletUTxO', balancedTx) -> do
-              _ <- swapTMVar utxoVar (walletUTxO', pparams, systemStart, epochInfo)
+              writeTVar utxoVar (walletUTxO', pparams, systemStart, epochInfo)
```

Comment:
Do we still want to do this? This is optimistically removing used utxos when balancing. But there is no guarantee that transaction will ever be included in the chain?

Reply:
Thinking out loud: the reason we would need this is because we want to prevent trying to double-spend some UTxO.  Double-spending would suppose that we try to submit more than one transaction at the time. 

- :heavy_check_mark: We can't commit until after Init has been observed
- :heavy_check_mark: We don't try to collect-com until after all commits have been observed
- :heavy_check_mark: We can't close until after collect-com has been observed
- :heavy_check_mark: We can't contest until after a close has been observed
- :x: We **might** try to fanout right after we submit a contest.

Indeed, we allow to contest right up until the deadline; which may lead in situation where the fanout is posted while a contest is pending. This is pretty bad because depending on how transactions gets re-organised due to chain fork, the fanout may end up being processed before / in stead of the contest. If we make sure that the utxo used for fanout depends on the one used for contest, then we at least limit this kind of issues. 

As a side-note, we wouldn't be able to fanout either because in normal scenarios, we have only one fuel UTxO. Thus, if a contest is pending, no UTxO is available and fanout becomes impossible. But I highly prefer failing for a "lack of payment input" than later having to handle a rejected transaction from the network because of a double-spending (which is precisely the kind of case we want to treat as _adversarial_ and for which there's no point retrying transactions). 

One thing we ought to implement IMO is _predicted UTxO_. Indeed, at this very moment, we know perfectly well what the transaction is and thus, what would be the resulting UTxO if included. We can thus, instead of removing the consumed UTxO, replaced it with the produced ones as if we had already observed the transaction. This doesn't change much the first 4 cases above, but it allows us to successfully submit a fanout on top of a pending contestation.

Reply:
Wait.. what will happen if a transaction fails to be submitted? e.g. the fanout because it was tried before the deadline? Will the UTxO be "made available" again? If not.. that is even a bug and we would *NOT* want to optimistically consume the UTxO at this location.

Reply:
Yup.

That's why we need validity intervals on transactions submitted as part of the Hydra protocol. Say 1h. If after 1h you haven't seen the transaction in a block, then you can legitimately "unlock" the UTxO that was "optimistically consume".

In the end, consuming the UTxO (as in, erasing it completely) is kind of a bad idea. Ideally we would keep two sets. One being the sets of UTxO currently part of pending transactions.

Reply:
Wouldn't it be a much simpler solution to just not bother and have the ledger sort out potential double spends? i.e. the corner case where two tx were balanced, signed and submitted using the same UTxO because of a race between observing the first tx and constructing the second tx?

After all, our own local cardano-node would not approve a double spend and if our application (or the client in proxy) wants to spam the chain with fanout txs it's our/their fault.

Reply:
I did not follow all the details, doing something else, but I would rather rely on retrying tx and some timeout based on actual observations than maintaining yet another state on-chain. If I understood correctly what's proposed. Or even better: Leave those details to a proper wallet implementation and just define some simpler "protocol" interaction between chain poster and attached wallet.